File: 2021/day14/day14.go
package main

import (
	"bufio"
	"log"
	"os"
	"strings"
)

func ReadFileToStr(fileName string) []string {
	file, err := os.Open(fileName)
	if err != nil {
		log.Fatalf("readLines: %s", err)
	}

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		val := scanner.Text()
		lines = append(lines, val)
	}
	return lines
}

func main() {
	lines := ReadFileToStr("day14.txt")
	startWord := lines[0]
	sequences := make(map[string]string)

	for _, line := range lines[2:] {
		s := strings.Split(line, " -> ")
		sequences[s[0]] = s[1]
	}
	log.Printf("Start word: %s", startWord)
	log.Printf("Sequences: %v", sequences)
}

File: 2021/day14/day14.py
import collections

with open("day14.txt", "r") as f:
    lines = [line.strip() for line in f.readlines()]

seq = lines[0]
rules = lines[2:]
rule_dict = {}
for rule in rules:
    rule_dict[rule[0:2]] = rule[6:]

def get_seq(iterations, seq):
    pairs_counter = collections.Counter()
    char_counter = collections.Counter()

    # Save intial chars to counter
    for c in seq:
        char_counter[c] += 1

    # Save intital pairs to counter
    for i in range(len(seq)-1):
        pairs_counter[seq[i:i+2]] += 1
        
    for _ in range(iterations):
        # Aux counter to keep track of new pairs this iteration
        new_pairs = collections.Counter()
        
        # Check for new pairings 
        for pair, v in pairs_counter.items():
            new_char = rule_dict[pair]

            # Add new pairings to counter            
            new_pairs[pair[0]+new_char] += v
            new_pairs[new_char+pair[1]] += v

            # Add the new character
            char_counter[new_char] += v

        pairs_counter = new_pairs
    
    occ = sorted(char_counter.values())
    return occ[-1] - occ[0]

print(f"Part 1: {get_seq(10, seq)}")
print(f"Part 2: {get_seq(40, seq)}")